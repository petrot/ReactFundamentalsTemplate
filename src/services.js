export const createUser = async (data) => {
  const response = await fetch("http://localhost:4000/register", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });

  return await response?.json();
};

export const login = async (data) => {
  const response = await fetch("http://localhost:4000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });

  return await response?.json();
};

export const getCourses = async () => {
  const response = await fetch("http://localhost:4000/courses/all", {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  return await response?.json();
};

export const getAuthors = async () => {
  const response = await fetch("http://localhost:4000/authors/all", {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  return await response?.json();
};

export const getCurrentUser = async (token) => {
  const response = await fetch("http://localhost:4000/users/me", {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: token,
    },
  });

  return await response?.json();
};

export const updateCourseService = async (data, token) => {
  const response = await fetch("http://localhost:4000/courses/" + data?.id, {
    method: "PUT",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
      Authorization: token,
    },
  });

  return await response?.json();
};

export const logout = async (token) => {
  const response = await fetch("http://localhost:4000/logout", {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      Authorization: token,
    },
  });

  return await response;
};

export const deleteCourseService = async (courseId, token) => {
  const response = await fetch("http://localhost:4000/courses/" + courseId, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      Authorization: token,
    },
  });

  return await response?.json();
};

export const createCourse = async (data, token) => {
  const response = await fetch("http://localhost:4000/courses/add", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
      Authorization: token,
    },
  });

  return await response?.json();
};

export const createAuthor = async (data, token) => {
  const response = await fetch("http://localhost:4000/authors/add", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
      Authorization: token,
    },
  });

  return await response?.json();
};
