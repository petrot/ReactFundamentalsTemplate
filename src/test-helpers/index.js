export * from "./local-storage.mock";
export * from "./store.mock";
export * from "./test-wrapper.mock";
