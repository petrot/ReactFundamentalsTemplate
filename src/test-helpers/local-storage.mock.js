export const mockGetItem = jest.fn();
export const mockSetItem = jest.fn();
export const mockRemoveItem = jest.fn();

export const prepareMockLocalStorage = () => {
  Object.defineProperty(window, "localStorage", {
    value: {
      getItem: (...args) => mockGetItem(...args),
      setItem: (...args) => mockSetItem(...args),
      removeItem: (...args) => mockRemoveItem(...args),
    },
  });
};
