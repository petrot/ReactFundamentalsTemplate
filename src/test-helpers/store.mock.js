export const TEST_NAME = "Test Name";

export const mockedState = {
  user: {
    isAuth: true,
    name: TEST_NAME,
    role: "admin",
    token: "mock-token",
  },
  courses: [
    {
      title: "Course 1",
      description: "Description 1",
      duration: 5,
      authors: [
        "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
        "1c972c52-3198-4098-b6f7-799b45903199",
        "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
      ],
      creationDate: "11/03/2024",
      id: "b28fd9ad-c683-4900-bf99-38a1c5748309",
    },
  ],
  authors: [
    {
      name: "author",
      id: "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
    },
    {
      name: "author2",
      id: "1c972c52-3198-4098-b6f7-799b45903199",
    },
    {
      name: "author3",
      id: "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
    },
  ],
};

export const mockedStore = {
  getState: () => mockedState,
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};
