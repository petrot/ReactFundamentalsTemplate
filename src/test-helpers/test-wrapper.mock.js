import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import { mockedStore } from "./store.mock";

export const TestWrapper = ({ children, store = mockedStore }) => (
  <Provider store={store}>
    <Router>{children}</Router>
  </Provider>
);
