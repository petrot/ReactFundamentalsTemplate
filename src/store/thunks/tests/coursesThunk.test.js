import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import * as Services from "../../../services";
import { mockedState } from "../../../test-helpers";
import {
  createCourseThunk,
  deleteCourseThunk,
  getCoursesThunk,
  updateCourseThunk,
} from "../coursesThunk";

const middlewares = [thunk];
const mockStoreCreator = configureStore(middlewares);

describe("[coursesThunk]", () => {
  let store;

  beforeEach(() => {
    store = mockStoreCreator();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should createCourseThunk dispatches saveCourse", async () => {
    const courseData = { ...mockedState.courses[0], title: "New course" };

    jest.spyOn(Services, "createCourse").mockResolvedValue({
      result: courseData,
    });

    await store.dispatch(createCourseThunk(courseData));
    const actions = store.getActions();

    expect(actions).toMatchSnapshot();
  });

  it("should updateCourseThunk dispatches updateCourse", async () => {
    const courseData = { ...mockedState.courses[0], title: "Modified" };

    jest.spyOn(Services, "updateCourseService").mockResolvedValue({
      result: courseData,
    });

    await store.dispatch(updateCourseThunk(courseData));
    const actions = store.getActions();

    expect(actions).toMatchSnapshot();
  });

  it("should updateCourseThunk not dispatches updateCourse without response id", async () => {
    const courseData = {
      ...mockedState.courses[0],
      title: "Modified",
      id: undefined,
    };

    jest.spyOn(Services, "updateCourseService").mockResolvedValue({
      result: courseData,
    });

    await store.dispatch(updateCourseThunk(courseData));
    const actions = store.getActions();

    expect(actions).toMatchInlineSnapshot(`Array []`);
  });

  it("should deleteCourseThunk dispatches deleteCourse", async () => {
    jest.spyOn(Services, "deleteCourseService").mockResolvedValue({
      result: "OK",
    });

    await store.dispatch(deleteCourseThunk("mockId"));
    const actions = store.getActions();

    expect(actions).toMatchSnapshot();
  });

  it("should deleteCourseThunk not dispatches deleteCourse without result", async () => {
    jest.spyOn(Services, "deleteCourseService").mockResolvedValue({
      result: undefined,
    });

    await store.dispatch(deleteCourseThunk("mockId"));
    const actions = store.getActions();

    expect(actions).toMatchInlineSnapshot(`Array []`);
  });

  it("should getCoursesThunk dispatches setCourses", async () => {
    jest
      .spyOn(Services, "getCourses")
      .mockResolvedValue({ result: mockedState.courses });

    await store.dispatch(getCoursesThunk());
    const actions = store.getActions();

    expect(actions).toMatchSnapshot();
  });
});
