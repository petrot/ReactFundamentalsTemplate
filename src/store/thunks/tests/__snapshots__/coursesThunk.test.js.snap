// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[coursesThunk] should createCourseThunk dispatches saveCourse 1`] = `
Array [
  Object {
    "payload": Object {
      "authors": Array [
        "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
        "1c972c52-3198-4098-b6f7-799b45903199",
        "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
      ],
      "creationDate": "11/03/2024",
      "description": "Description 1",
      "duration": 5,
      "id": "b28fd9ad-c683-4900-bf99-38a1c5748309",
      "title": "New course",
    },
    "type": "courses/saveCourse",
  },
]
`;

exports[`[coursesThunk] should deleteCourseThunk dispatches deleteCourse 1`] = `
Array [
  Object {
    "payload": "mockId",
    "type": "courses/deleteCourse",
  },
]
`;

exports[`[coursesThunk] should getCoursesThunk dispatches setCourses 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "authors": Array [
          "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
          "1c972c52-3198-4098-b6f7-799b45903199",
          "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
        ],
        "creationDate": "11/03/2024",
        "description": "Description 1",
        "duration": 5,
        "id": "b28fd9ad-c683-4900-bf99-38a1c5748309",
        "title": "Course 1",
      },
    ],
    "type": "courses/setCourses",
  },
]
`;

exports[`[coursesThunk] should updateCourseThunk dispatches updateCourse 1`] = `
Array [
  Object {
    "payload": Object {
      "authors": Array [
        "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
        "1c972c52-3198-4098-b6f7-799b45903199",
        "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
      ],
      "creationDate": "11/03/2024",
      "description": "Description 1",
      "duration": 5,
      "id": "b28fd9ad-c683-4900-bf99-38a1c5748309",
      "title": "Modified",
    },
    "type": "courses/updateCourse",
  },
]
`;
